web3.js/src/connection.ts

```diff
@@ -2761,80 +2762,257 @@ export class Connection {
       try {
-        subscriptionId = this.onSignature(
-          signature,
-          (result: SignatureResult, context: Context) => {
-            subscriptionId = undefined;
-            response = {
-              context,
-              value: result,
-            };
-            resolve(null);
-          },
-          subscriptionCommitment,
+        await promiseTimeout(confirmPromise, timeoutMs);
+      } finally {
+        if (subscriptionId) {
+          this.removeSignatureListener(subscriptionId);
+        }
+      }
+
+      if (response === null) {
+        const duration = (Date.now() - start) / 1000;
+        throw new Error(
+          `Transaction was not confirmed in ${duration.toFixed(
+            2,
+          )} seconds. It is unknown if it succeeded or failed. Check signature ${signature} using the Solana Explorer or CLI tools.`,
         );
+      }
+      return response;
+    } else {
+      try {
+        decodedSignature = bs58.decode(signature.signature);
       } catch (err) {
-        reject(err);
+        throw new Error(
+          'signature must be base58 encoded: ' + signature.signature,
+        );
       }
-    });
 
-    let timeoutMs = this._confirmTransactionInitialTimeout || 60 * 1000;
-    switch (subscriptionCommitment) {
-      case 'processed':
-      case 'recent':
-      case 'single':
-      case 'confirmed':
-      case 'singleGossip': {
-        timeoutMs = this._confirmTransactionInitialTimeout || 30 * 1000;
-        break;
-      }
-      // exhaust enums to ensure full coverage
-      case 'finalized':
-      case 'max':
-      case 'root':
-    }
+      assert(decodedSignature.length === 64, 'signature has invalid length');
+
+      const subscriptionCommitment = commitment || this.commitment;
+      let subscriptionId;
+      let done = false;
+      let response: RpcResponseAndContext<SignatureResult> | null = null;
+
+      const confirmTx = new Promise((resolve, reject) => {
+        try {
+          subscriptionId = this.onSignature(
+            signature.signature,
+            (result: SignatureResult, context: Context) => {
+              subscriptionId = undefined;
+              response = {
+                context,
+                value: result,
+              };
+              done = true;
+              resolve(TransactionStatus.CONFIRMED);
+            },
+            subscriptionCommitment,
+          );
+        } catch (err) {
+          reject(err);
+        }
+      });
 
-    try {
-      await promiseTimeout(confirmPromise, timeoutMs);
-    } finally {
-      if (subscriptionId) {
-        this.removeSignatureListener(subscriptionId);
+      const checkBlockHeight = async () => {
+        const blockHeight = await this.getBlockHeight(commitment);
+        return blockHeight;
+      };
+
+      const expireTx = new Promise((resolve, reject) => {
+        (async () => {
+          try {
+            let currentBlockHeight = await checkBlockHeight();
```

Comment:
You might be done by the time this finishes.
```suggestion
            let currentBlockHeight = await checkBlockHeight();
            if (done) return;
```

Reply:
Thanks, I'll adjust this. I knew this could be improved somehow!